explorer/interpreter/type_checker.cpp

```diff
@@ -381,7 +381,9 @@ static auto TypeContainsAuto(Nonnull<const Value*> type) -> bool {
     case Value::Kind::PointerType:
       return TypeContainsAuto(&cast<PointerType>(type)->pointee_type());
     case Value::Kind::StaticArrayType:
-      return TypeContainsAuto(&cast<StaticArrayType>(type)->element_type());
+      const auto* array_type = cast<StaticArrayType>(type);
```

Comment:
To me it seems like a great point.
Maybe something like `DeduceableType` or `IncompleteType` would do the work?

Another inconsistency with the naming that I've stumbled on is the usage of the `concrete` word in the context of types and in the context of inheritance. E.g. `IsConcreteType` means "type without things to deduce" and `ExpectConcreteType` means "expect non-abstract class". I believe, I've already seen similar discussion somewhere on Discord, but not sure that it had come to some conclusion.

Reply:
> To me it seems like a great point. Maybe something like `DeduceableType` or `IncompleteType` would do the work?

Actually, looking at the implementation again, maybe `TypeIsDeduceable` because it asserts it's a type, versus returning false if it's not a type.

"Incomplete" is a little different: e.g., we call a class as incomplete during evaluation of its body. I think `class C;` would declare an incomplete class.

> Another inconsistency with the naming that I've stumbled on is the usage of the `concrete` word in the context of types and in the context of inheritance. E.g. `IsConcreteType` means "type without things to deduce" and `ExpectConcreteType` means "expect non-abstract class". I believe, I've already seen similar discussion somewhere on Discord, but not sure that it had come to some conclusion.

I look at this as slight drift over time. Maybe we should rename `IsConcreteType` to `IsNonDeduceableType` to be clearer? I could do that in a separate PR if you prefer.

Reply:
`TypeIsDeduceable` sounds good to me, as well as `IsNonDeduceableType`.

I think I could address both renamings in this PR, if it's convenient, or in the separate PR, after we're done with this one.